File: system/monolith/dotnet/Core/Dtos/PublishCouponRequest.cs
using System.ComponentModel.DataAnnotations;

namespace Optivem.Shop.Monolith.Core.Dtos;

public class PublishCouponRequest : IValidatableObject
{
    [Required(ErrorMessage = "Coupon code must not be blank")]
    public string? Code { get; set; }

    [Required(ErrorMessage = "Discount rate must not be null")]
    public decimal? DiscountRate { get; set; }

    public DateTime? ValidFrom { get; set; }

    public DateTime? ValidTo { get; set; }

    [Range(1, int.MaxValue, ErrorMessage = "Usage limit must be positive")]
    public int? UsageLimit { get; set; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (DiscountRate.HasValue)
        {
            if (DiscountRate.Value <= 0m)
            {
                yield return new ValidationResult("Discount rate must be greater than 0.00", new[] { nameof(DiscountRate) });
            }
            else if (DiscountRate.Value > 1.0m)
            {
                yield return new ValidationResult("Discount rate must be at most 1.00", new[] { nameof(DiscountRate) });
            }
        }
    }
}

File: system/multitier/backend-java/src/main/java/com/mycompany/myshop/backend/core/dtos/PublishCouponRequest.java
package com.mycompany.myshop.backend.core.dtos;

import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.Data;

import java.math.BigDecimal;
import java.time.Instant;

@Data
public class PublishCouponRequest {
    @NotBlank(message = "Coupon code must not be blank")
    private String code;

    @NotNull(message = "Discount rate must not be null")
    @DecimalMin(value = "0.0", inclusive = false, message = "Discount rate must be greater than 0.00")
    @DecimalMax(value = "1.0", message = "Discount rate must be at most 1.00")
    private BigDecimal discountRate;

    private Instant validFrom;

    private Instant validTo;

    @Positive(message = "Usage limit must be positive")
    private Integer usageLimit;
}
